### PR TITLE
Attempt to shutdown plugin correctly

### PR DESCRIPTION
**Pull request by garethjevans**
To validate:

```
TEST=api make test-autoscaler-suite
```

and monitor with:

```
ps -ef | grep cred-helper | grep -v ps | grep -v grep
```

### CODE REVIEW

File: src/autoscaler/cred_helper/loader.go [hunk 1]
```diff
@@ -15,9 +16,23 @@ import (
 
 type PluginManager struct {
 	client *plugin.Client
+
+	// ctx is used to shutdown the device manager
+	ctx    context.Context
```
Comment (KevinJCross) [inline src/autoscaler/cred_helper/loader.go:21]: do we use this context anywhere ?
we should probably  be using it in the RPC client requests ?

Comment (KevinJCross) [reply src/autoscaler/cred_helper/loader.go:21]: OK ... so https://www.quora.com/What-is-the-difference-between-kill-command-and-interrupt-a-process
suggests that TERM should be caught and kill(9) should not be called.
So we should be calling sig term or `.Terminate()` and killwithFire should not be used (SEG_F/SIGKILL)

File: src/autoscaler/metricsforwarder/cmd/metricsforwarder/main.go [hunk 2]
```diff
@@ -64,13 +64,14 @@ func main() {
 	}
 	defer policyDB.Close()
 
-	pm := cred_helper.PluginManager{}
+	pm := cred_helper.New()
+	defer pm.Kill()
```
Comment (KevinJCross) [inline src/autoscaler/metricsforwarder/cmd/metricsforwarder/main.go:68]: hmmm ... with the terminate does the plugin close the db connection it creates ?
Actually we should be doing this in the defer too.

File: src/autoscaler/api/cmd/api/api_test.go [hunk 3]
```diff
@@ -28,6 +28,7 @@ var _ = Describe("Api", func() {
 	})
 
 	AfterEach(func() {
+		// then destroy
 		runner.KillWithFire()
```
[the remark this answers is not in the document]
Comment (KevinJCross) [reply src/autoscaler/api/cmd/api/api_test.go:32]: in sigmon.New it adds TERM and Interupt
```

func New(runner ifrit.Runner, signals ...os.Signal) ifrit.Runner {
	signals = append(signals, syscall.SIGINT, syscall.SIGTERM)
	return &sigmon{
		Signals: signals,
		Runner:  runner,
	}
}
```

So we can use either interrupt or term intead of `KillWithFire` ... typically Ive used the kill -9 when the process gets stuck terminating ... so we should attempt to do 
`Terminate()` Poll for termination for x seconds and if this does not terminate we should then call `kill -9` or `Kill()`

**Comment by KevinJCross** [comment]
(x) Change needed. We need to replace `KillWithFIre()` withn preferably Terminate() or Interrupt().
Because `KillWithFIre()` terminates immediately and does not allow gracefull shutdown (closing connections) and should only be used in frozen or broken processes

**Comment by garethjevans** [comment]
@KevinJCross the terminate change looks good, but I can't approve my own PR

File: src/autoscaler/api/cmd/api/main.go [hunk 4]
```diff
@@ -80,13 +80,14 @@ func main() {
 		os.Exit(1)
 	}
 
-	pm := cred_helper.PluginManager{}
+	pm := cred_helper.New()
```
Comment (asalan316) [inline src/autoscaler/api/cmd/api/main.go:83]: Is the plugin manager needed although we have only one plugin here?

**Review by KevinJCross — APPROVED** [review]

**Review by asalan316 — APPROVED** [review]

**Pull request merged**